src/execution/join_hashtable.cpp

```diff
@@ -910,16 +926,125 @@ idx_t ScanStructure::ResolvePredicates(DataChunk &keys, SelectionVector &match_s
 		result_count = this->count;
 	}
 
+	if (ht.residual_predicate && result_count > 0) {
+		result_count = ApplyResidualPredicate(probe_data, match_sel, result_count, no_match_sel);
+	}
+
 	// Update total probe match count
 	ht.total_probe_matches.fetch_add(result_count, std::memory_order_relaxed);
 
 	return result_count;
 }
 
-idx_t ScanStructure::ScanInnerJoin(DataChunk &keys, SelectionVector &result_vector) {
+idx_t ScanStructure::ApplyResidualPredicate(DataChunk &probe_data, SelectionVector &match_sel, idx_t match_count,
+                                            SelectionVector *no_match_sel) {
+	if (match_count == 0 || !ht.residual_predicate) {
+		return match_count;
+	}
+
+	// determine total columns needed based on ORIGINAL indices
+	idx_t total_columns = 0;
+
+	// find max probe column index (original position)
+	for (const auto &entry : ht.probe_input_to_probe_map) {
+		total_columns = MaxValue(total_columns, entry.first + 1);
+	}
+
+	// find max build column index (with offset)
+	for (const auto &entry : ht.build_input_to_layout_map) {
+		total_columns = MaxValue(total_columns, entry.first + 1);
+	}
+
+	// build evaluation chunk structure
+	vector<LogicalType> eval_types(total_columns, LogicalType::INVALID);
+	vector<bool> initialize_columns(total_columns, false);
+
+	// fill in probe types at their ORIGINAL positions
+	for (const auto &entry : ht.probe_input_to_probe_map) {
+		idx_t orig_idx = entry.first;
+		idx_t probe_data_col = entry.second;
+
+		eval_types[orig_idx] = probe_data.data[probe_data_col].GetType();
+		initialize_columns[orig_idx] = true;
+	}
+
+	// fill in build types at their offset positions
+	for (const auto &entry : ht.build_input_to_layout_map) {
+		idx_t col_with_offset = entry.first;
+		idx_t layout_col = entry.second;
+
+		eval_types[col_with_offset] = ht.layout_ptr->GetTypes()[layout_col];
+		initialize_columns[col_with_offset] = true;
+	}
+
+	// initialize evaluation chunk with selective initialization
+	DataChunk eval_chunk;
+	eval_chunk.Initialize(Allocator::Get(ht.context), eval_types, initialize_columns, STANDARD_VECTOR_SIZE);
+	eval_chunk.SetCardinality(match_count);
+
+	// copy probe columns at their ORIGINAL positions
+	for (const auto &entry : ht.probe_input_to_probe_map) {
+		idx_t orig_idx = entry.first;
+		idx_t probe_data_col = entry.second;
+
+		eval_chunk.data[orig_idx].Slice(probe_data.data[probe_data_col], match_sel, match_count);
+	}
+
+	// gather RHS columns from hash table
+	for (const auto &entry : ht.build_input_to_layout_map) {
+		idx_t col_with_offset = entry.first;
+		idx_t layout_col = entry.second;
+
+		auto &target_vector = eval_chunk.data[col_with_offset];
+		GatherResult(target_vector, match_sel, match_count, layout_col);
+	}
+
+	// execute the residual predicate
+	DataChunk result_chunk;
+	result_chunk.Initialize(ht.context, {LogicalType::BOOLEAN});
+	residual_executor->Execute(eval_chunk, result_chunk);
```

Comment:
Instead of initializing these chunks here each time on the fly, could we store them in `JoinHashTable::ProbeState` instead? This allows us to just initialize them once, and `DataChunk::Reset()` them each time we use them. If you put them as `unique_ptr<ResidualPredicateProbeState>`, you can optionally initialize the chunks. This should be more efficient and clean up this function